src/v/cluster/partition_balancer_planner.cc

```diff
@@ -0,0 +1,333 @@
+/*
+ * Copyright 2022 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "cluster/partition_balancer_planner.h"
+
+#include "cluster/health_monitor_types.h"
+#include "cluster/scheduling/constraints.h"
+#include "cluster/scheduling/types.h"
+#include "cluster/types.h"
+#include "model/fundamental.h"
+#include "model/metadata.h"
+
+namespace cluster {
+
+partition_balancer_planner::partition_balancer_planner(
+  topic_table& topic_table,
+  partition_allocator& partition_allocator,
+  planner_config config)
+  : _config(config)
+  , _topic_table(topic_table)
+  , _partition_allocator(partition_allocator) {}
+
+void partition_balancer_planner::
+  init_ntp_sizes_and_node_reports_from_health_report(
+    const cluster_health_report& health_report) {
+    for (const auto& node_report : health_report.node_reports) {
+        _node_reports[node_report.id] = node_report;
+        for (const auto& tp_ns : node_report.topics) {
+            for (const auto& partition : tp_ns.partitions) {
+                _ntp_sizes[model::ntp(
+                  tp_ns.tp_ns.ns, tp_ns.tp_ns.tp, partition.id)]
+                  = partition.size_bytes;
+            }
+        }
+    }
+    return;
+}
+
+bool partition_balancer_planner::check_partition_movement_is_available(
+  const std::vector<model::broker_shard>& current_replicas) {
+    // Check that nodes quorum is available
+    size_t available_nodes_amount = std::count_if(
+      current_replicas.begin(),
+      current_replicas.end(),
+      [this](const model::broker_shard& bs) {
+          return _unavailable_nodes.find(bs.node_id)
+                 == _unavailable_nodes.end();
+      });
+    if (available_nodes_amount * 2 < current_replicas.size()) {
+        return false;
+    }
+    return true;
+}
+
+size_t partition_balancer_planner::get_partition_size(
+  model::ns n, model::topic t, model::partition_id p) {
+    const auto ntp_data = _ntp_sizes.find(model::ntp(n, t, p));
+    if (ntp_data == _ntp_sizes.end()) {
+        vlog(
+          clusterlog.error,
+          "Partition {} {} {} status was not found in cluster health report",
+          n,
+          t,
+          p);
+    } else {
+        return ntp_data->second;
+    }
+    return size_t(-1);
```

Comment:
if you have to use special values why not wrap in `result<>`

Reply:
Changed to constexpr value